proposals/0000-implicit-resource-binding.md

```diff
@@ -0,0 +1,369 @@
+resource. This is called _implicit resource binding_ and it is the focus of this
+document.
+
+## Current behavior in DXC
+
+This section documents the implicit resource binding assignment in DXC. For
+simplicty the examples below use the just the `u` registers and
+`RWBuffer<float>` resource, but same binding assignment rules apply to other
+register and resource types as well.
+
+### Simple Resource Declarations
+
+Resource without binding assignment that are declared directly at the global
+scope are processed in the order they appear in the source code. They are
+assigned the first available register slot in `space0`. Unused resources are
+optimized away and do not participate in the implicit binding assignment.
+
+#### Example 1.1
+```c++
+RWBuffer<float> A : register(u0);
+RWBuffer<float> B;                 // gets u1
+RWBuffer<float> C: register(u2);   // unused
+RWBuffer<float> D;                 // gets u2
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[0] + B[0];
+}
+```
+https://godbolt.org/z/sMs3c8j7T
+
+If `C` is used then `D` gets assigned descriptor `u3`. 
+
+However, if the resources are declared inside a `struct`, their binding seems be
+assigned in the order they are referenced in the code, and not the order in
+which they were declared:
+
+#### Example 1.2
+```c++
+RWBuffer<float> A : register(u0);
+RWBuffer<float> C: register(u2);   // unused
+
+struct S {
+  RWBuffer<float> B;                 // gets u2
+  RWBuffer<float> D;                 // gets u1
+} s;
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = s.D[0] + s.B[0];
+}
+```
+https://godbolt.org/z/6a1b7Kvz3
+
+The maximum value of the register slot and virtual space number is `UINT_MAX`
+(max. value of 32-unsigned integer).
+
+### Constant-size Resource Arrays
+
+Constant-size resource arrays declared directly at the global scope are also
+processed in the order they appear in the source code. They are assigned the
+first range of available decriptors in `space0` that fits the array size
+regardless of whether the individual resources in the array are used or not, as
+long as at least one of the them is accessed.
+
+If a resource isn't used, the whole resource declaration is optimized away and
+its descriptor range is available for use by other resources.
+
+#### Example 2.1
+```c++
+RWBuffer<float> A : register(u2);
+RWBuffer<float> B[4];    // gets u3 because it does not fit before A (range 4)
+RWBuffer<float> C[2];    // gets u0 because it fits before A (range 2)
+RWBuffer<float> D[50] : register(u6); // unused
+RWBuffer<float> E[2];    // gets u7 which is right after B (range 4)
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = E[2][0] + C[1][0] + B[1][0];
+}
+```
+https://godbolt.org/z/qha84sjvT
+
+However, if the resource array is defined in a `struct`, the binding seems to be
+assigned to the individual array elements in the order they are used in the
+code. In other words, the array elements are treated as unrelated individual
```

Comment:
If the substitution for the word `code` that I made in my comment above is true, you might want to replace `code` too.
And if you can determine whether this behavior is solid or not, you may want to remove `seems to be`.